src/crimson/os/seastore/journal.cc

```diff
@@ -384,4 +242,498 @@ Journal::replay_ret Journal::replay(
+  OrderingHandle& handle)
+{
+  assert(p_current_batch->is_empty());
+  record_batch_stats.increment(1);
+  auto total = rsize.mdlength + rsize.dlength;
+  return seastar::now().then([this, total] {
+    if (journal_segment_manager.needs_roll(total)) {
+      return journal_segment_manager.roll();
+    } else {
+      return JournalSegmentManager::roll_ertr::now();
+    }
+  }).safe_then([this, rsize, record=std::move(record), &handle]() mutable {
+    increment_io();
+    ceph::bufferlist to_write = encode_record(
+      rsize,
+      std::move(record),
+      journal_segment_manager.get_block_size(),
+      journal_segment_manager.get_committed_to(),
+      journal_segment_manager.get_nonce());
+    auto write_fut = journal_segment_manager.write(to_write);
+    return handle.enter(write_pipeline->device_submission
+    ).then([write_fut=std::move(write_fut)]() mutable {
+      return std::move(write_fut);
+    }).finally([this] {
+      decrement_io_with_flush();
+    }).safe_then([this, &handle, rsize](journal_seq_t write_start) {
+      return mark_record_committed_in_order(handle, write_start, rsize);
+    });
+  });
+}
+
+Journal::RecordSubmitter::submit_pending_ret
+Journal::RecordSubmitter::submit_pending(
+  record_t&& record,
+  const record_size_t& rsize,
+  OrderingHandle& handle,
+  bool flush)
+{
+  assert(!p_current_batch->is_submitting());
+  record_batch_stats.increment(
+      p_current_batch->get_num_records() + 1);
+  auto write_fut = p_current_batch->add_pending(
+    std::move(record), rsize);
+  if (flush) {
+    flush_current_batch();
+  }
+  return handle.enter(write_pipeline->device_submission
+  ).then([write_fut=std::move(write_fut)]() mutable {
+    return std::move(write_fut);
+  }).safe_then([this, &handle, rsize](journal_seq_t write_start) {
+    return mark_record_committed_in_order(handle, write_start, rsize);
+  });
+}
+
+Journal::RecordSubmitter::do_submit_ret
+Journal::RecordSubmitter::do_submit(
+  record_t&& record,
+  const record_size_t& rsize,
+  OrderingHandle& handle)
+{
+  assert(!p_current_batch->is_submitting());
+  if (state <= state_t::PENDING) {
+    assert(!wait_submit_promise.has_value());
+    if (p_current_batch->is_pending()) {
+      auto batched_size = p_current_batch->get_encoded_length(rsize);
+      if (batched_size > journal_segment_manager.get_max_write_length()) {
+        flush_current_batch();
+        return do_submit(std::move(record), rsize, handle);
+      } else if (journal_segment_manager.needs_roll(batched_size)) {
+        flush_current_batch();
+        return journal_segment_manager.roll(
+        ).safe_then([this, record=std::move(record), rsize, &handle]() mutable {
+          return do_submit(std::move(record), rsize, handle);
+        });
+      } else {
+        return submit_pending(std::move(record), rsize, handle, true);
+      }
+    } else {
+      assert(p_current_batch->is_empty());
+      return submit_record(std::move(record), rsize, handle);
```

Comment:
What's the difference between submit_record and submit_pending(..., true)?

Reply:
If they are essentially the same, we can eliminate submit_record entirely and remove the is_pending() branch here.

Reply:
I think the basic rule is that records should be persistent in the same order of submission, this implies that the pending batch needs to be submitted before the current record.

So, if `p_current_batch->is_pending()` is false, `RecordSubmitter` can safely submit the current record directly in `submit_record()` without `RecordBatch`.

If `is_pending()` is true, and if the current record is **safe** to be add to `p_current_batch`, `RecordSubmitter` will add the current record to the end of the current `RecordBatch` and flush the batch together, which is the `submit_pending(..., true)` branch.

Reply:
Right, it's that first case I'm not so sure of.  Is it actually measurably more expensive to construct a single record RecordBatch?  This isn't an obviously worthwhile optimization to me given the extra code.

Reply:
I think the fast path saves a `RecordBatch` with a shared-promise (potentially with seastar scheduling) and with some vectors.

I will append a separate commit removing the fast-path and see what's the difference.

Reply:
We're allocating RecordBatch's up front in Journal::batches and reusing them, right?  We can eliminate the std::vector growth allocations by giving each RecordBatch a fixed size std::array.  The shared-promise might matter, but it seems like we'd probably want to address that directly rather than adding a second fast path we need to maintain.

Reply:
Yeah, this way looks reasonable to me.

Reply:
> We can eliminate the std::vector growth allocations by giving each RecordBatch a fixed size std::array. The shared-promise might matter, but it seems like we'd probably want to address that directly rather than adding a second fast path we need to maintain.

All addressed, see https://github.com/ceph/ceph/pull/43617#issuecomment-958690945